### ISSUE TITLE
Add option to auto-save answer

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#3358]
Students writing long answers to exam questions may be unable to save due to running out of time or network connectivity issues. There should be an option to automatically save the submission periodically via AJAX.

Workaround is to add to top of question:

```
<script>
  let serial = false;
  function saveSubmission() {
    let new_serial = $('.question-form').serialize();
    if (serial !== new_serial) {
      serial = new_serial;
      $.post(window.location.href, serial + '&__action=save');
    }
  }
  setInterval(saveSubmission, 10000);
</script>
```

Which will save every 10 seconds if something changed, but it will create several submissions. Ideally this should replace an existing submission instead.

**Comment by mwest1066** [comment]
> Which will save every 10 seconds if something changed, but it will create several submissions. Ideally this should replace an existing submission instead.

Yep, I agree that we don't want to build up thousands of submissions because someone left a tab open 😄 

The thing I've always struggled with about autosaving is that it seems hard to get it really robust. I guess we'd need to frequently push in the background (on every change? rate limited to once per second?) and then somehow indicate to the user when we failed to push? I guess we could have a "saving..." indicator in the corner to show we are trying to push. The retry logic around failed pushes makes me a bit nervous.

But I guess it wouldn't be too bad to have an ajax call pushing submission state in the background, and the server will always update the latest submission with this new data. We'd just need to be careful with the notifications and retry logic.

Although I'm also worried about the "Next/prev question" buttons. These do a full page reload, so we'd lost any autosave pushes running in the background. I'm not sure how to handle this robustly. Maybe we just notify and don't otherwise worry about it?

**Comment by jonatanschroeder** [comment]
One of the reasons I'm asking this is because some of our students take courses using other platforms that have autosave, so they may rely on this behaviour.  Anyway, some ideas:

For avoiding too many saves, as in the suggested code, we can compare the serialization of the form, and only save it different.

As for the submission, my suggestion is a new flag in the submission that indicates it's autosaved. Any future submission from any kind (autosaved, regular save or grade) would replace this submission if one exists.

Next/prev could maintain their current behaviour of warning if the submission is not saved. This comparison could then be done with the last autosaved submission that returned successfully.

**Comment by firasm** [comment]
I'm teaching a writing course this summer where students have to write short essays in short periods of time and this would be really important factor in deciding whether to use PL for the course or not.

**Comment by aroaryan** [comment]
I think it would particularly help at-least in exam conditions when a student would be writing at the end moment. Just a save at the end.

**Comment by trombonekenny** [comment]
This came up again this week:

> I recently took an exam for CS 173. However, as I was answering the last question, which was a written response question that had a blue save button to click in order to save your written answer, I wasn't able to hit the save button before the exam time ran out, so I was wondering if anything I wrote for that question saved. And if it did not save, is there anything that could be done?

**Comment by margaretmfleck** [comment]
> One of the reasons I'm asking this is because some of our students take courses using other platforms that have autosave, so they may rely on this behaviour. 

And, conversely, experience with a platform that loses answers may cause students to worry unnecessarily when taking exams on a system (e.g. Moodle) that saves aggressively.    

For long-answer questions, moodle autosave just protects the answer currently being written.  So it doesn't create multiple submissions.   There are entirely other pathways that can create multiple submissions for the same long-answer question on moodle.   So I can tell you for free that this is VERY BAD when instructors/staff are trying to do manual grading.    The only thing worse than slogging through grading proofs is discovering that some of your work is going into grading answers that have been supplanted by newer ones.

**Comment by echuber2** [comment]
In the short term, a workaround might be to make it a workspace question instead. If students author a markdown document in Jupyter, that can autosave. The autosave interval can be tweaked in the workspace settings (maybe it would be appropriate to customize this in a derived image beforehand, if that's the usecase -- also to open a named markdown template file by default instead of a notebook or something).

**Comment by margaretmfleck** [comment]
If we did that, would we still get the rest of the grading interface for open-answer questions, esp. model answer and rubric?

**Comment by echuber2** [comment]
I'm not sure how manual grading goes along with workspaces (haven't used it yet). @jonatanschroeder Do you know?

**Comment by echuber2** [comment]
What if the assignment auto-submitted once _at the deadline_, if not more often? Then you could choose to only grade the submission without opening the workspace, but the workspace would still offload the autosaving logic while the student works.

Separately, I wonder how many negative side effects there would be to strictly controlling access to existing workspaces based on the access rules. If it's problematic, maybe the synced directory could be frozen (remounted read-only?) if there is no active access rule for submission. Or the other concept of just relying on an auto-submission to save one last copy before the deadline might have fewer negative side effects.

**Comment by jonatanschroeder** [comment]
> What if the assignment auto-submitted once _at the deadline_, if not more often? Then you could choose to only grade the submission without opening the workspace, but the workspace would still offload the autosaving logic while the student works.

That is what I assume you mean. What I'm saying is that, if the assessment is available (even if inactive or closed), the student can modify the workspace, so using the workspace as a workaround for autosave has its caveats. There is a way to handle it, such as fully block access to the assessment while manual grading is in place (which many instructors do anyway), which would block access to the workspace as well (I think), but that adds some limitations.

> Separately, I wonder how many negative side effects there would be to strictly controlling access to existing workspaces based on the access rules.

This is already done (afaik), though there may be caveats if access ends while the workspace is already open (I remember this being an issue in the past, I don't remember if this was handled).

> If it's problematic, maybe the synced directory could be frozen (remounted read-only?) if there is no active access rule for submission. 

I believe that may not be enough.

> Or the other concept of just relying on an auto-submission to save one last copy before the deadline might have fewer negative side effects.

The issue is: when do you save that one last copy? What if network is interrupted at that point? What if the submission takes longer than the cut-off to be saved (e.g., if there is network delay or congestion)?

**Comment by echuber2** [comment]
I thought workspace submissions were collected from the synced file storage, which is in the cloud, not the client side. So there should be some assurance about being able to queue or retry a scheduled final collection of their cloud contents even if the student's client connection is unstable. It's true there may still be an issue with a student continuing to manually edit or submit approximately _at_ the deadline itself but it doesn't seem like a showstopper. If network congestion in the cloud prevents the submission from being performed for several minutes or longer after the deadline, and there's no CBTF involved, students may be able to sneak in a last tweak; but what's the greater concern for the particular instructor then?

**Comment by reteps** [comment]
An alternative solution would be to still require the student to manually save their answer, but also continually snapshot their progress. Then, at a later point, if they want to "Save from snapshot" they can. Then they are still explicitly saving, but they can still recover their work if something goes wrong.

**Comment by margaretmfleck** [comment]
It's important to remember that any of these systems can suffer a hard crash, not just a controlled timeout at the end of the exam.    This is especially true for in-class assessments where we're working with student laptops.   But a badly-aimed backhoe can likely bring down the PL server or its connection to the rest of the world.   Also, as we recently saw with the DFA-building tools, small bugs in an element can cause work to be erased when the student pushes "save" or "save and grade".    In that case, the student did nothing at all wrong and yet the work is gone.

In my experience, students are most distressed by the "PL ate my work" scenarios, and don't feel that it's really their fault that they weren't constantly pushing "save".    They are ok with the idea that they might lose a couple words at the end of an answer if there is a system crash, and graders can usually work around that sort of glitch.   Although I'm sure they'd like a history feature in case they accidentally delete work themselves and then have regrets, they do understand that this situation is their own fault.  When this happened on paper, only the most chill faculty members were willing to try to decipher erased text to salvage some points.

Having grown up in an earlier time, I automatically save editor buffers every paragraph or so.  But that's very last century.

**Comment by echuber2** [comment]
(I'd like to highlight Pete's last post because GitHub doesn't seem to send a notification about edits. A substantial edit to a comment might be better posted as a followup comment.)

> Having grown up in an earlier time, I automatically save editor buffers every paragraph or so. But that's very last century.

I also have the habit of doing Ctrl-S constantly, but I guess that hotkey would conflict with browser page "save as."

It seems like automatically doing the save periodically, or upon navigation, maybe in a background Shared Worker thread, could work? I also think the workspace approach can work in some cases.

**Comment by jonatanschroeder** [comment]
> It seems like automatically doing the save periodically, or upon navigation, maybe in a background Shared Worker thread, could work?

Note that this was the original intent of the redacted code above, however this brings some issues to consider:

* Depending on the kind of question and the frequency of updates, sending a full submission periodically involves a significant amount of network traffic (which is bad both for students and the infrastructure), especially if the questions involve file editors or uploads. A proper implementation should be aware of when changes occur and only upload changes if there was an update.
* The issue above is even trickier when workspaces are involved.
* Should every periodic update create a new submission? That might cause a problem with significant amount of pollution in the student screen, as well as confusion for graders if there is an expectation of giving credit to previous submissions. Ideally a proper implementation of this feature should account for that and save the automated updates either in a separate data structure or update an existing submission as changes arise (i.e., a "provisional" submission of sorts).
    * Of course this brings up the issue of snapshots: while a submission may be created when a student explicitly calls "Save", should there be additional saved snapshots that allow either a student or instructor to visualize the saved state in between submissions? Is there a value to that?
* Ideally the update should only send deltas, not full submissions, to reduce the network impact. This might require some elements (e.g., file editor and rich text editor) to have some notion of deltas that can be extracted client-side and interpreted server-side to combine the deltas to an existing submission. This is also tricky when considering workspace-based submissions.
* The frequency of updates should also be discussed. Too frequent and you end up with overuse of resources. Too few and you end up not saving all the data.
* Potential issues of temporary interruptions should also be taken into account. What happens if network connectivity is lost for some time? Especially, what happens if this is close to the end of the exam period?
* Does this need to be done in homeworks, or exams with laxer time periods (e.g., >24 hours)?

**Comment by margaretmfleck** [comment]
You folks should seriously consider investigating how this is handled by other systems.   It would be hard to find out what's happening on the back end with, say, google docs.   But Moodle is open source and manages this part smoothly.    UIUC (at least) has a moodle support group who might just know the answers or (at least) who to ask.

What I can tell you from the instructor side of Moodle is that they save aggressively.   When the browser front-end crashes, students never lose more than a word or two.    However, when I look at someone's exam after it closes, Moodle shows me only selected milestones, only slightly more detailed than what I see on PL's history display.    So apparently most of its saves are kept only temporarily.   

As you say, they might be sending deltas.   But, seriously, ask those folks rather than trying to guess.   Even if you don't want to do the exact same thing, it's likely to be informative to see a previous design.   

Notice that I've frequently been hitting UIUC's campus Moodle server with 200-400 students taking a quiz in parallel without any noticeable responsiveness issues.   

I would strongly suggest not creating extra submissions, beyond what students explicitly save/submit and an autosave at the end.   The main need is to enable that final autosave.   Secondarily to allow students to recover from a crash midway through an assessment.     It's rare to want to dig into earlier submissions when grading.

**Comment by reteps** [comment]
I looked at some Moodle API calls, it looked like they were autosaving long-form text inputs every 45 seconds, using the full submission. It only makes that request if content changes on the input. I didn't see any autosave for checkbox/MCQ questions. The autosave was also element-based, so it would do separate autosaves per field.

**Comment by echuber2** [comment]
I think it would be better to avoid creating a homebrew diff system and just submit the full data each time, possibly letting transmission libraries automatically optimize that based on a binary delta like `rsync` does. If that still proves to be too bandwidth-intensive, maybe shorter-interval saves should happen in localStorage only. Or, use a two-stage system of saving to localStorage and then transmitting the same data as an autosave (with delta transfer enabled).

**Comment by nwalters512** [comment]
I think we should shift discussion to concrete proposals. Here's two as starting points. I'm not saying these are perfect or even good, but I'm hoping that dealing in specifics will help us move forward.

## Option 1: generic auto-save with server-side persistence

We'll introduce a new `pending_submissions` table with the following columns:

- `id`
- `date`
- `variant_id`
- `raw_submitted_answer`
- A `UNIQUE` constraint on `variant_id`

On the client, we'll do a background submission of form data (possibly only changed data, I think it probably doesn't really matter) on some regular schedule, for the sake of discussion maybe this:

- After 10 seconds of unchanged form values
- At least every minute

This will upsert a row in `pending_submissions`.

When the user loads a variant, we'll check for a `pending_submission` for the current variant with a `date` greater than the date of the latest submission. If such a pending submission is found, we'll show a message to the user like "Unsaved work found. Click here to restore it." If the user opts to do so, we'll basically just call [`processSubmission()`](https://github.com/PrairieLearn/PrairieLearn/blob/f81fbaf495b65156d9f0667a82fc5b0399d822ab/apps/prairielearn/src/lib/question-submission.ts#L10-L19) with `raw_submitted_answer` used where `req.body` usually would be, and then re-render the page.

> [!NOTE]
> This is necessary because basically all elements assume that if they're rendered for the `question` panel with a given submission, the data in `data["submitted_answers"]` and friends has already been processed by `parse()`, which is always run as part of saving a submission. There's no way for us to just splat raw form field values into the form.
>
> The downside of this is that there's no great way for students to preview the saved data before persisting it as a (saved, not graded) submission. IMO this is acceptable. _However_, we could in theory show a preview by doing a `parse()` + `render()` on the fly without actually persisting the resulting submission anywhere.

We'd probably implement some kind of auto-pruning here, e.g. remove any pending submissions that were created more than 7 days ago.

In a trivial implementation, this information wouldn't be visible to instructors. That is, it would still be the responsibility of students to click "Save" before a timer runs out. We _could_ make this information visible to instructors on the manual grading page by doing a `parse()` + `render()` on the fly without actually persisting the resulting submission anywhere. As an enhancement, we could give instructors the ability to commit the pending submission as a real submission there.

## Option 2: element-specific client-side persistence

We can just shove things into `localStorage` on the client. This would mainly guard against browser crashes, computer power loss, and things like that. We'd limit this to elements where students are likely to build up a lot of complex state, e.g. `<pl-file-editor>`, `<pl-rich-text-editor>`, `<pl-excalidraw>`, etc. This would be simpler overall, but would require a lot of duplicate boilerplate across both first-party and third-party elements. We'd also have to deal with evicting old data, as `localStorage` is limited to ~5MB in most browsers. It's possible that we could do this in a centralized way, though it's more likely that this would be boilerplate we'd have to copy around to different elements.

**Comment by reteps** [comment]
I think students would most benefit from a system that lets them preview their submissions over time. Could option 1 be implemented in a way similar to `submission_panels`, that lets them see their submission after it has been parsed and rendered?

This is roughly what I am envisioning; similar to option 1. They preview the "submission" state. You still persist the `raw_submitted_answer`. I also like the term snapshot over `pending_submission` -- i feel like this implies that pending submissions are all wiped from the database upon a save.

![Image](https://github.com/user-attachments/assets/3459b73c-d0f5-4a5e-b08d-f8ced689b1b8)

**Comment by jonatanschroeder** [comment]
One issue that is not considered in either of @nwalters512's solutions above, and it's a major concern in this issue (it's in the first sentence of the issue description), is what happens if a student runs out of time and loses access to the assessment before the timer runs out. In that case, who would have the authority to apply these pending submissions, and what information would they have regarding if that submission should be considered or not (e.g., maybe the student was drafting something out but it was to be ignored)? Should students be allowed to trigger that pending submission to be committed after the timer has run out (considering that the pending submission data was created before the timer ran out)?

Another question: what happens if a student switches between questions without saving? Should we allow a student to have more than one pending submission in the same assessment?

> we'll check for a `pending_submission` for the current variant with a `date` greater than the date of the latest submission.

An alternative is to have the submission insertion process delete any pending submissions, then we don't need to check the date. This also reduces the need for purging old data.

**Comment by reteps** [comment]
My view is that if no saved submission is present on a question, then the instructor should have access to the snapshot history (maybe just the last one), and can manually award points. I was also thinking that pending submissions would be per question.

> An alternative is to have the submission insertion process delete any pending submissions, then we don't need to check the date. This also reduces the need for purging old data.

I think this is a fine way of doing this, the other way would be to only keep the most recent X snapshots.

**Comment by jonatanschroeder** [comment]
> My view is that if no saved submission is present on a question, then the instructor should have access to the snapshot history (maybe just the last one), and can manually award points.

This would require changing the manual grading to also list instance questions with pending submissions and no proper submissions. Currently instance questions without a submission are not listed. This is not a major concern, just listing it to make sure it's considered in the implementation.

As for "award points", I'm assuming that the instructor would then "commit" the pending submission (i.e., convert it to a proper submission, which would run parse and potentially grade), since we need the submission record to keep track of things like grading jobs, feedback messages and rubrics.

A question remains, though, in automatically graded questions: should the autograder (internal or external) run in such submissions? If so, should the instructor trigger it or should it be done automatically? Note that in many of these cases an instructor expects the submissions to be graded automatically, so the existence of any pending submission would need to be presented to the instructor in a visible manner outside the manual grading interface.

**Comment by nwalters512** [comment]
> I also like the term snapshot over pending_submission -- i feel like this implies that pending submissions are all wiped from the database upon a save.

In my proposal, they more or less are: note the `UNIQUE` constraint and the fact that we'd only show a pending submission if it's newer than the most recent submission. Maintaining a full history would explode our database usage for very little value if we don't do some kind of very fancy delta saving, and I'd very much rather avoid that.

> An alternative is to have the submission insertion process delete any pending submissions, then we don't need to check the date. This also reduces the need for purging old data.

This could work as well if we're careful, yes.

> In that case, who would have the authority to apply these pending submissions, and what information would they have regarding if that submission should be considered or not (e.g., maybe the student was drafting something out but it was to be ignored)? Should students be allowed to trigger that pending submission to be committed after the timer has run out (considering that the pending submission data was created before the timer ran out)?

IMO we leave this all to instructors via manual grading. The alternative is too tricky. Consider an exam with real-time grading disabled that times out with a pending submission. The assessment likely would have been immediately graded when it closed. If the student were able to opt to "promote" specific pending submissions, what's the implication for grading? I'd rather we don't even attempt to handle that and leave this as a very manual escape hatch for both students and instructors.

> As for "award points", I'm assuming that the instructor would then "commit" the pending submission (i.e., convert it to a proper submission, which would run parse and potentially grade), since we need the submission record to keep track of things like grading jobs, feedback messages and rubrics.

I'm thinking the same thing, yeah.

> A question remains, though, in automatically graded questions: should the autograder (internal or external) run in such submissions? If so, should the instructor trigger it or should it be done automatically?

IMO we certainly shouldn't try to grade it automatically, mainly because guessing which `credit` value should apply at the moment it's graded would be tricky. Maybe we let instructors opt to do it manually, and give them control over credit at the time they "commit" the pending submission?

> Note that in many of these cases an instructor expects the submissions to be graded automatically, so the existence of any pending submission would need to be presented to the instructor in a visible manner outside the manual grading interface.

Maybe, though I really truly view this as a last-resort edge-case kind of feature. The norm should continue to be that students save (or save+grade); IMO autosaving is for a) browser crashes and b) extenuating circumstances. Tell me if you think I'm thinking about this the wrong way though!

**Comment by echuber2** [comment]
If instructors manually review the pending (un-)autograded submission, is there a mechanism for manually triggering the autograder to run as a convenience? I thought this was a longstanding feature request for regrades.

About the storage size limitation, maybe IndexedDB could be used instead of localStorage. There is still a limit but more breathing room.

**Comment by reteps** [comment]
> Maintaining a full history would explode our database usage for very little value if we don't do some kind of very fancy delta saving, and I'd very much rather avoid that.

I'm proposing partial history here. In https://github.com/PrairieLearn/PrairieLearn/issues/3358#issuecomment-2833636913 I mention this scenario:

> The other scenario is if the student saves a answer, and then starts rewriting it, but cannot finish because they are out of time. In this case, the student likely wants to submit their original answer, and not the half-finished rewrite that they didn't save. I wouldn't want any sort of autosaving feature to overwrite what the student has explicitly saved with an autosaved version.

The student would likely have to go back later than one autosave to recover their submission. If it autosaves every 10 seconds, this might not work well. I would propose an autosave of 30 seconds or when you leave the page. Then you store the last n=10 submissions, or 5 minutes of history.

**Comment by nwalters512** [comment]
> The other scenario is if the student saves a answer, and then starts rewriting it, but cannot finish because they are out of time. In this case, the student likely wants to submit their original answer, and not the half-finished rewrite that they didn't save. I wouldn't want any sort of autosaving feature to overwrite what the student has explicitly saved with an autosaved version.

I assumed this was a given based on how I had imagined this working. As in, auto-save doesn't actually create a real submission; it's always still on the student to press "Save" to commit it (or communicate to the instructor that XYZ disaster happened and can they please consider the pending submission).

Put differently: in the above scenario, the original answer was _already_ saved, and the student can opt to use it by simply not saving anything after it, the same as things are today.

This is one of the reasons I don't think we should try to elevate pending submissions in the UI somehow: I don't think an instructor can know if a given submission can or should be considered without considering the student's input and other factors.

**Comment by jonatanschroeder** [comment]
I would prefer that partial snapshots be shifted to a different issue. The concern here is that something is saved, not that we keep track of past snapshots. We currently have a mechanism for a manual snapshot: "Save only". If there is a use-case for a partial snapshot between submissions I think that's a different problem than what we have yere.

**Comment by margaretmfleck** [comment]
How do you distinguish a student "rewriting" their answer from a student who has saved the first section of their long answer and is now partway through finishing their answer?    They would want us to see everything they managed to get onto "paper". not merely the beginning section where they took a moment to save.

**Comment by jonatanschroeder** [comment]
> How do you distinguish a student "rewriting" their answer from a student who has saved the first section of their long answer and is now partway through finishing their answer? They would want us to see everything they managed to get onto "paper". not merely the beginning section where they took a moment to save.

I think this is where the merit of @nwalters512's suggestion of handling this manually comes into play: the instructor takes the decision.

**Comment by reteps** [comment]
This was discussed in today's meeting; with the combination of undo and the last submission, the usefulness of multiple snapshots seems marginal. However, it might be nice to have an implementation that tracks the last N=... snapshots where N=1 right now if this is something we want to revisit.
